subsys/lorawan/nvm/lorawan_nvm_settings.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2021 Intellinium <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <LoRaMac.h>
+#include <settings/settings.h>
+#include "lorawan_nvm.h"
+#include <logging/log.h>
+#if defined(CONFIG_LORAWAN_SE_HAS_OWN_NVM)
+#include "../se/lorawan_se.h"
+#endif
+
+LOG_MODULE_REGISTER(lorawan_nvm, CONFIG_LORAWAN_LOG_LEVEL);
+
+static uint16_t nvm_notify_flag;
+
+#define LORAWAN_SETTINGS_BASE                "lorawan/nvm"
+
+#define flag_set(f)        ((nvm_notify_flag & (f)) == (f))
+
+#define SAVE_SETTINGS(nvm, flag, name)					\
+if (flag_set(flag)) {							\
+	LOG_DBG("Saving configuration " LORAWAN_SETTINGS_BASE "/" # name);\
+	int err = settings_save_one(LORAWAN_SETTINGS_BASE "/" # name,	\
+				&(nvm)->name,				\
+				sizeof((nvm)->name));			\
+	if (err) {							\
+		LOG_ERR("Could not save settings" # name", error %d",	\
+			err);						\
+	}								\
+}
+
+static void nvm_settings_work(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	MibRequestConfirm_t mibReq;
+
+	LOG_DBG("Saving LoRaWAN settings");
+
+	/* Retrieve the actual context */
+	mibReq.Type = MIB_NVM_CTXS;
+	if (LoRaMacMibGetRequestConfirm(&mibReq) != LORAMAC_STATUS_OK) {
+		LOG_ERR("Could not get NVM context");
+		return;
+	}
+
+	LoRaMacNvmData_t *nvm = mibReq.Param.Contexts;
+
+	LOG_DBG("Crypto version: %"PRIu32", DevNonce: %d, JoinNonce: %"PRIu32,
+		mibReq.Param.Contexts->Crypto.LrWanVersion.Value,
+		mibReq.Param.Contexts->Crypto.DevNonce,
+		mibReq.Param.Contexts->Crypto.JoinNonce);
+
+	/* Wait for LoRa to be idle */
+	while (LoRaMacStop() != LORAMAC_STATUS_OK) {
```

Comment:
Why does the MAC layer need to be stopped and restarted here?

Reply:
This is taken from the loramac-node [sample app](https://github.com/zephyrproject-rtos/loramac-node/blob/zephyr/src/apps/LoRaMac/common/NvmDataMgmt.c#L72).
I guess it is to prevent more IRQ to be fired which would also change the NVM values at the same time, inducing race condition.
However, I'm not an expert in the LoRaMAC layer, so that could be unnecessary.